libs/community/langchain_community/vectorstores/sqlserver.py

```diff
@@ -141,18 +148,25 @@ def _insert_embeddings(
             List of IDs generated from adding the texts into the vectorstore.
         """
 
-        if ids is None:
-            ids = [uuid.uuid4() for _ in texts]
-
         if metadatas is None:
             metadatas = [{} for _ in texts]
 
+        if ids is None:
+            ids = [metadata.pop("id", uuid.uuid4()) for metadata in metadatas]
+
         try:
             with Session(bind=self._bind) as session:
                 documents = []
-                for id, query, embedding, metadata in zip(
-                    ids, texts, embeddings, metadatas
-                ):
+                for idx, query in enumerate(texts):
+                    # For a query, if there is no corresponding ID, we generate a uuid and add it to the list of IDs to be returned.
+                    id = (
+                        ids[idx]
+                        if idx < len(ids)
+                        else ids.append(uuid.uuid4()) or ids[-1]
```

Comment:
why is there an `or` here? when do we get the last member - ids[-1]?

Reply:
When list of IDs is less than list of text and we iterate to a text that does not have an `id`, we create a new ID and append it to the list of IDs and then assign that appended value to variable `id` (this is done by `ids[-1]`).  The expression `ids.append(uuid.uuid4())` will always evaluate to `false`, and `ids[-1]` evaluates to `true`. The `or` operator allows `ids[-1]` to be executed after the append expression.